presto-raptor/src/test/java/com/facebook/presto/raptor/storage/TestOrcFileRewriter.java

```diff
@@ -413,9 +428,225 @@ public void testUncompressedSize(boolean useOptimizedOrcWriter)
+        File newFile2 = new File(temporary, randomUUID().toString());
+        info = createFileRewriter(useOptimizedOrcWriter).rewrite(
+                getColumnTypes(ImmutableList.of(7L, 10L), ImmutableList.of(createVarcharType(20), DOUBLE)),
+                newFile1,
+                newFile2,
+                new BitSet(5));
+        assertEquals(info.getRowCount(), 1);
+
+        // Add a column with the different ID with different type
+        File newFile3 = new File(temporary, randomUUID().toString());
+        info = createFileRewriter(useOptimizedOrcWriter).rewrite(
+                getColumnTypes(ImmutableList.of(7L, 10L, 13L), ImmutableList.of(createVarcharType(20), DOUBLE, createVarcharType(5))),
+                newFile2,
+                newFile3,
+                new BitSet(5));
+        assertEquals(info.getRowCount(), 1);
+
+        // Get prepared for the final file; make sure it is accessible from storage manager
+        UUID uuid = randomUUID();
+        File newFile4 = getFileSystemPath(new File(dataDir, "data/storage"), uuid);
+        if (useOptimizedOrcWriter) {
+            // Apache ORC writer creates the file itself
+            newFile4.getParentFile().mkdirs();
+            newFile4.createNewFile();
+        }
+
+        // Drop a column and add a column
+        info = createFileRewriter(useOptimizedOrcWriter).rewrite(
+                getColumnTypes(ImmutableList.of(7L, 13L, 8L), ImmutableList.of(createVarcharType(20), createVarcharType(5), INTEGER)),
+                newFile3,
+                newFile4,
+                new BitSet(5));
+        assertEquals(info.getRowCount(), 1);
+
+        ConnectorPageSource source = storageManager.getPageSource(
+                uuid,
+                OptionalInt.empty(),
+                ImmutableList.of(13L, 7L, 8L),
+                ImmutableList.of(createVarcharType(5), createVarcharType(20), INTEGER),
+                getColumnTypes(ImmutableList.of(13L, 7L, 8L), ImmutableList.of(createVarcharType(5), createVarcharType(20), INTEGER)),
+                TupleDomain.all(),
+                READER_ATTRIBUTES);
+
+        Page page = null;
+        while (page == null) {
+            page = source.getNextPage();
+        }
+        assertEquals(page.getPositionCount(), 1);
+
+        // Column 13L
+        Block column0 = page.getBlock(0);
+        assertTrue(column0.isNull(0));
+
+        // Column 7L
+        Block column1 = page.getBlock(1);
+        assertEquals(createVarcharType(20).getSlice(column1, 0), utf8Slice("2"));
+
+        // Column 8L
+        Block column2 = page.getBlock(2);
+        assertTrue(column2.isNull(0));
+
+        dummyHandle.close();
+        deleteRecursively(dataDir.toPath(), ALLOW_INSECURE);
+    }
+
+    /**
+     * The following test drop and add the same columns; the legacy ORC rewriter will fail due to unchanged schema.
+     * However, if we enforce the newly added column to always have the largest ID, this won't happen.
```

Comment:
I think that's what we do in the current code base but we don't have a good way to enforce it. See my comments above on the possible fix.